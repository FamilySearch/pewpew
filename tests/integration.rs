use std::env;

use futures::FutureExt;
use test_common::{start_test_server, TestWriter};
use tokio::runtime::Runtime;

fn run_test(path: &str) -> (bool, String, String) {
    let mut rt = Runtime::new().unwrap();
    rt.block_on(async move {
        let (port, kill_server, _) = start_test_server(None);
        env::set_var("PORT", port.to_string());

        let (_, ctrlc_channel) = futures::channel::mpsc::unbounded();

        let run_config = pewpew::RunConfig {
            config_file: path.into(),
            output_format: pewpew::RunOutputFormat::Human,
            results_dir: Some("./".into()),
            stats_file: "integration.json".into(),
            stats_file_format: pewpew::StatsFileFormat::Json,
            start_at: None,
            watch_config_file: true,
        };
        let exec_config = pewpew::ExecConfig::Run(run_config);

        let stdout = TestWriter::new();
        let stderr = TestWriter::new();

        let stdout2 = stdout.clone();
        let stderr2 = stderr.clone();

        let success = pewpew::create_run(exec_config, ctrlc_channel, stdout, stderr)
            .map(|r| if r.is_ok() { true } else { false })
            .await;

        let _ = kill_server.send(());

        (success, stdout2.get_string(), stderr2.get_string())
    })
}

#[test]
fn int1() {
    let (success, _stdin, stderr) = run_test("tests/integration.yaml");

    assert!(success, "test run failed. {}", stderr);

    let left = stderr;
    let right = include_str!("integration.stderr.out");
    assert_eq!(left, right);
}

#[test]
fn int_on_demand() {
    let (success, _stdin, stderr) = run_test("tests/int_on_demand.yaml");

    assert!(success, "test run failed. {}", stderr);

    assert!(
        stderr.len() > 0,
        "expected stderr to be a bunch of '1'. Instead saw: {}",
        stderr
    );

    for line in stderr.lines() {
        assert_eq!(
            line, "1",
            "expected stderr to be a bunch of '1'. Instead saw: {}",
            stderr
        );
    }
}
