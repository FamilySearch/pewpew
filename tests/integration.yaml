load_pattern:
  - !linear
      to: 100%
      over: 5s
config:
  client:
    headers:
      test: 123
      float: 1.23
loggers:
  l:
    limit: 2
    to: !stderr
  timeLogger:
    to: !file 'test-epoch-${x:epoch("ms")}.json'
vars:
  b: "foo"
  e2: ["foo", "foo", "foo"]
  port: "${e:PORT}"
providers:
  a:
    !range
  c:
    !list
      - 123
      - 456
  d:
    !response
  f:
    !file
      path: integration.data
      repeat: true
endpoints:
  - method: POST
    declare:
      e: !c
        collects:
          - take: 3
            from: '${v:b}'
            as: _b
        then: ${p:_b}
    url: http://localhost:${v:port}/
    body: !str '{"a": ${p:a}, "b": "${v:b}", "c": ${p:c}, "e": ${p:e}, "e2": ${v:e2}, "f": "${p:f}"}'
    peak_load: 1.1hps
    headers:
      Content-Type: application/json
    logs:
      l:
        select:
          each:
            - 'response.body.b == "foo"'
            - 'val_eq(response.body.e, response.body.e2)'
            - 'request.headers.test == "123"'
            - 'parseInt(request.headers.test) + 1 == 124'
            - 'parseInt(request.headers.float) == 1'
            - 'parseFloat(request.headers.float) == 1.23'
            - 'parseInt(response.body.b) == null'
            - 'response.body.a == 0'
            - 'response.body.c == 123'
            - 'response.body.f == "line1"'
            - 'response.body.a == 1'
            - 'response.body.c == 456'
            - 'response.body.f == "line2"'
          all: >-
            (response.body.b == "foo"
              && val_eq(response.body.e, response.body.e2)
              && request.headers.test == "123"
              && parseInt(request.headers.test) + 1 == 124
              && parseInt(request.headers.float) == 1
              && parseFloat(request.headers.float) == 1.23
              && parseInt(response.body.b) == null
            )
            &&
            (
              (response.body.a == 0
                && response.body.c == 123
                && response.body.f == "line1")
              ||
              (response.body.a == 1
                && response.body.c == 456
                && response.body.f == "line2")
            )
