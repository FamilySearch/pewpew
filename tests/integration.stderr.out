{"all":true,"each":[true,true,true,true,true,true,true,true,true,true,false,false,false]}
{"all":true,"each":[true,true,true,true,true,true,true,false,false,false,true,true,true]}
